File: ruby/lib/runapi/topaz/resources/image_upscales.rb
# frozen_string_literal: true

module RunApi
  module Topaz
    module Resources
      class ImageUpscales
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/topaz/image_upscales"

        RESPONSE_CLASS = Types::ImageUpscaleResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedImageUpscaleResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "model is required" unless param(params, :model) == Types::IMAGE_UPSCALE_MODEL
          raise Core::ValidationError, "image_url is required" unless param(params, :image_url)

          factor = param(params, :upscale_factor)
          raise Core::ValidationError, "upscale_factor is required" unless factor
          return if Types::IMAGE_UPSCALE_FACTORS.include?(factor.to_s)

          raise Core::ValidationError, "upscale_factor must be one of: #{Types::IMAGE_UPSCALE_FACTORS.join(", ")}"
        end
      end
    end
  end
end

File: js/src/resources/image-upscales.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { ImageUpscaleParams, ImageUpscaleResponse, TaskCreateResponse } from '../types';

const ENDPOINT = '/api/v1/topaz/image_upscales';

export class ImageUpscales {
  constructor(private readonly http: HttpClient) {}

  async run(params: ImageUpscaleParams, options?: RequestOptions & PollingOptions): Promise<ImageUpscaleResponse> {
    const { id } = await this.create(params, options);
    return pollUntilComplete<ImageUpscaleResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
  }

  async create(params: ImageUpscaleParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<ImageUpscaleResponse> {
    return this.http.request<ImageUpscaleResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}
